weather_dl/download_pipeline/clients.py

```diff
@@ -152,6 +170,90 @@ def __exit__(self, exc_type, exc_value, traceback):
         self._redirector.__exit__(exc_type, exc_value, traceback)
 
 
+class APIRequestExtended(api.APIRequest):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def fetch(self, request):
+        status = None
+
+        self.connection.submit("%s/%s/requests" % (self.url, self.service), request)
+        self.log("Request submitted")
+        self.log("Request id: " + self.connection.last.get("name"))
+        if self.connection.status != status:
+            status = self.connection.status
+            self.log("Request is %s" % (status,))
+
+        while not self.connection.ready():
+            if self.connection.status != status:
+                status = self.connection.status
+                self.log("Request is %s" % (status,))
+            self.connection.wait()
+
+        if self.connection.status != status:
+            status = self.connection.status
+            self.log("Request is %s" % (status,))
+
+        result = self.connection.result()
+        return result
+
+    def download(self, result, target=None):
+        if target:
+            if os.path.exists(target):
+                # Empty the target file, if it already exists, otherwise the
+                # transfer below might be fooled into thinking we're resuming
+                # an interrupted download.
+                open(target, "w").close()
+
+            size = -1
+            tries = 0
+            while size != result["size"] and tries < 10:
+                size = self._transfer(
+                    urljoin(self.url, result["href"]), target, result["size"]
+                )
+                if size != result["size"] and tries < 10:
+                    tries += 1
+                    self.log("Transfer interrupted, resuming in 60s...")
+                    time.sleep(60)
+                else:
+                    break
+
+            assert size == result["size"]
+
+        self.connection.cleanup()
+
+        return result
+
+
+class ECMWFServiceExtended(ECMWFService):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def fetch(self, req):
```

Comment:
If it's not too much trouble: Could you add types to these two function signatures?

Reply:
Consider: creating type aliases for the return value (a named alias to `t.Dict` could be helpful for using types to document the feature).